functions.php

```diff
@@ -959,6 +964,105 @@ private function render_partial( $path, $args = [] ) {
 		}
 		include( locate_template( $path . '.php' ) );
 	}
+
+	/**
+	 * Add p4_page_type structure in available permalink tags for Settings -> Permalinks page.
+	 * available_permalink_structure_tags filter.
+	 *
+	 * @param $tags
+	 *
+	 * @return mixed
+	 */
+	public function add_p4_page_type_permalink_structure( $tags ) {
+		$tags['p4_page_type'] = __( 'P4 page type (A p4 page type term.)' );
+
+		return $tags;
+	}
+
+	/**
+	 * Replace p4_page_type placeholder with the p4_page_type term for posts permalinks.
+	 * Filter for post_link.
+	 *
+	 * @param $permalink
+	 * @param $post_id
+	 * @param $leavename
+	 *
+	 * @return mixed
+	 */
+	public function p4_page_type_permalink( $permalink, $post_id, $leavename ) {
```

Comment:
Methods of a class that has prefixed name and its own namespace do not need to have prefixed names as well.